core/app/models/spree/product.rb

```diff
@@ -230,8 +230,14 @@ def generate_slug
         having("SUM(#{Spree::StockItem.table_name}.count_on_hand) <= 0")
     }
     scope :out_of_stock, -> { joins(:stock_items).where("#{Spree::StockItem.table_name}.count_on_hand <= ? OR #{Spree::Variant.table_name}.track_inventory = ?", 0, false) }
-    scope :single_variant, lambda {
-      where(variants_count: 0)
+
+    scope :by_best_selling, lambda { |order_direction = :desc|
```

Comment:
@damianlegawiec I replaced queries with using counter caches with this scope (same for classifications)